packages/flame/test/components/component_lifecycle_test.dart

```diff
@@ -35,74 +38,111 @@ void main() {
+      await game.add(parent);
+      await game.ready();
 
       // The parent tries to prepare the component before it is added to the
       // game and fails since it doesn't have a game root and therefore re-adds
       // the child when it has a proper root.
-      expect(
-        parentEvents,
-        ['onGameResize', 'prepared: true', 'onLoad', 'onMount'],
-      );
-      expect(
-        childEvents,
-        [
-          'prepared: false',
-          'onGameResize',
-          'prepared: true',
-          'onLoad',
-          'onMount',
-        ],
-      );
+      expect(parentEvents, ['onGameResize', 'onLoad', 'onMount']);
+      expect(childEvents, ['onGameResize', 'onLoad', 'onMount']);
     });
 
     flameGame.test('correct lifecycle on parent change', (game) async {
       final parentEvents = <String>[];
       final childEvents = <String>[];
-      final parent = _MyComponent(parentEvents);
-      final child = _MyComponent(childEvents);
+      final parent = _MyComponent(parentEvents, 'parent');
+      final child = _MyComponent(childEvents, 'child');
       await parent.add(child);
       await game.ensureAdd(parent);
       child.changeParent(game);
       game.update(0);
+      await game.ready();
 
-      expect(
-        parentEvents,
-        [
-          'onGameResize',
-          'prepared: true',
-          'onLoad',
-          'onMount',
-        ],
-      );
+      expect(parentEvents, ['onGameResize', 'onLoad', 'onMount']);
       // onLoad should only be called the first time that the component is
       // loaded.
       expect(
         childEvents,
         [
-          'prepared: false',
           'onGameResize',
-          'prepared: true',
           'onLoad',
           'onMount',
+          'onRemove',
           'onGameResize',
-          'prepared: true',
           'onMount',
         ],
       );
     });
+
+    flameGame.test(
+      'components added in correct order even with different load times',
+      (game) async {
+        final a = SlowComponent(0.1);
+        final b = SlowComponent(0.02);
+        final c = SlowComponent(0.05);
+        final d = SlowComponent(0);
+        game.add(a);
+        game.add(b);
+        game.add(c);
+        game.add(d);
+        await game.ready();
+        expect(game.children.toList(), equals([a, b, c, d]));
+      },
+    );
+
+    test(
+      'Component starts loading before the parent is mounted',
+      () async {
+        final game = SingletonGame()
+          ..onGameResize(Vector2.all(100))
+          ..onMount();
+        final parent = Component();
+        final child = SlowComponent(0.01);
+        final future = child.addToParent(parent);
+        expect(parent.isMounted, false);
+        expect(parent.isLoaded, false);
+        expect(child.isMounted, false);
+        expect(child.isLoaded, false); // not yet..
+        await future;
+        expect(parent.isMounted, false);
+        expect(child.isLoaded, true);
```

Comment:
How can the child be loaded before the branch has been added to the root?

Reply:
This particular test uses a game with `SingleGameInstance` mixin, which preserves the eager loading mechanics.